### PR TITLE
Colored chat for FreeType builds

### PR DESCRIPTION
**Pull request by nerzhul**
From @triblade9 #2136 rebased and maintained PR (for dev-0.5 branch)

From the original version:
- Moved and rename ColoredString and StaticText to util/coloredstring and util/statictext

### CODE REVIEW

**Comment by kilbith** [comment]
Preview : https://lut.im/BMFWzvM9/RSJMPPNE

**Comment by ShadowNinja** [comment]
This is missing a reset token, and shouldn't it be merged to master since it's non-breaking?

**Comment by indriApollo** [comment]
Yes love u nerzhul ! :3

File: src/chat.cpp [hunk 1]
```diff
@@ -677,6 +680,14 @@ ChatBackend::~ChatBackend()
 void ChatBackend::addMessage(std::wstring name, std::wstring text)
 {
 	// Note: A message may consist of multiple lines, for example the MOTD.
+
+	// Strip colors out of chat if not using freetype.
+	#if USE_FREETYPE
+	#else
```
Comment (ShadowNinja) [inline src/chat.cpp]: Just use `#if !USE_FREETYPE`.

File: src/chat.cpp [hunk 2]
```diff
@@ -726,7 +737,11 @@ std::wstring ChatBackend::getRecentChat()
 	{
 		const ChatLine& line = m_recent_buffer.getLine(i);
 		if (i != 0)
-			stream << L"\n";
+			#if USE_FREETYPE
+				stream << L"\n\vffffff";
```
Comment (ShadowNinja) [inline src/chat.cpp]: Shouldn't auto-reset on every line like this IMO.

File: src/cguittfont/CGUITTFont.cpp [hunk 3]
```diff
@@ -654,7 +664,20 @@ void CGUITTFont::draw(const core::stringw& text, const core::rect<s32>& position
...
+				if (i < applied_colors.size())
+					col = applied_colors[i];
+			} else {
+				col = irr::video::SColor(255, 255, 255, 255);
+			}
+			if (!use_transparency)
+					col.color |= 0xff000000;
+			Driver->draw2DImage(page->texture, page->render_positions[i], page->render_source_rects[i], clip, col, true);
+		}
+		// Driver->draw2DImageBatch(page->texture, page->render_positions, page->render_source_rects, clip, color, true);
```
Comment (ShadowNinja) [inline src/cguittfont/CGUITTFont.cpp]: This should be removed, not just commented out.

File: src/cguittfont/CGUITTFont.cpp [hunk 4]
```diff
@@ -654,7 +664,20 @@ void CGUITTFont::draw(const core::stringw& text, const core::rect<s32>& position
 			for (size_t i = 0; i < page->render_positions.size(); ++i)
 				page->render_positions[i] -= core::vector2di(shadow_offset, shadow_offset);
 		}
-		Driver->draw2DImageBatch(page->texture, page->render_positions, page->render_source_rects, clip, color, true);
+		for (size_t i = 0; i < page->render_positions.size(); ++i) {
+			irr::video::SColor col;
+			if(applied_colors.size() > 0) {
+				col = applied_colors[0];
+				if (i < applied_colors.size())
+					col = applied_colors[i];
```
Comment (ShadowNinja) [inline src/cguittfont/CGUITTFont.cpp]: This should only do a lookup once, like:
```C++

if (applied_colors.size() > 0) {
	col = i < applied_colors.size() ? applied_colors[i] : col = applied_colors[0];
} else {
	col = irr::video::SColor(255, 255, 255, 255);
}
```

File: src/util/string.cpp [hunk 5]
```diff
@@ -613,8 +613,42 @@ static bool parseNamedColorString(const std::string &value, video::SColor &color
...
+	while (i < s.length()) {
+		if (s[i] == L'\v') {
+			i += 7;
+			continue;
+		}
+		output += s[i];
+		++i;
+	}
+	return output;
+}
```
Comment (ShadowNinja) [inline src/util/string.cpp]: * `// Removes escape sequences` is misleading.  It removes color escape sequences.
  * This should have a more descriptive name, like `removeChatColors`.
  * Passing `"\v"` or similar will make this crash (accessing out-of-bounds string index).

File: doc/lua_api.txt [hunk 6]
```diff
@@ -1826,6 +1826,10 @@ and `minetest.auth_reload` call the authetification handler.
 ### Chat
 * `minetest.chat_send_all(text)`
 * `minetest.chat_send_player(name, text)`
+* `minetest.colorize(color, text)`
+	  * `color` is a six characters string, this will reset color to white after the text
+*  `minetest.color(color)`
+    * Returns a valid escape sequence for `color`
```
Comment (ShadowNinja) [inline doc/lua_api.txt]: The values under these are indented inconsistently.

File: src/game.cpp [hunk 7]
```diff
@@ -2096,11 +2100,21 @@ bool Game::initGui(std::wstring *error_message)
...
 
 	// Chat text
+
+#if USE_FREETYPE
+	//Colored chat support when using FreeType
+	guitext_chat = new gui::StaticText(L"", false, guienv, guiroot, -1, core::rect<s32>(0, 0, 0, 0), false);
+	guitext_chat->setWordWrap(true);
+	guitext_chat->drop();
+#else
+	//Standard chat when FreeType is disabled
```
Comment (ShadowNinja) [inline src/game.cpp]: Missing space after `//` in comments.

File: src/guiChatConsole.cpp [hunk 8]
```diff
@@ -317,13 +317,27 @@ void GUIChatConsole::drawText()
...
+				irr::gui::CGUITTFont *tmp = static_cast<irr::gui::CGUITTFont*>(m_font);
+				tmp->draw(
+					fragment.text.c_str(),
+					destrect,
+					fragment.text.getColors(),
+					false,
+					false,
+					&AbsoluteClippingRect);
+			#else
+			//Otherwise use standard text
```
Comment (ShadowNinja) [inline src/guiChatConsole.cpp]: Missing space after `//` in comments.

File: src/util/CMakeLists.txt [hunk 9]
```diff
@@ -1,5 +1,14 @@
+if(USE_FREETYPE)
+	set(UTIL_freetypedep_SRCS
```
Comment (ShadowNinja) [inline src/util/CMakeLists.txt]: Why the weird capitalization?

File: src/util/CMakeLists.txt [hunk 10]
```diff
@@ -7,4 +16,8 @@ set(UTIL_SRCS
 	${CMAKE_CURRENT_SOURCE_DIR}/sha1.cpp
 	${CMAKE_CURRENT_SOURCE_DIR}/string.cpp
 	${CMAKE_CURRENT_SOURCE_DIR}/timetaker.cpp
+	${UTIL_freetypedep_SRCS}
 	PARENT_SCOPE)
+
+
+
```
Comment (ShadowNinja) [inline src/util/CMakeLists.txt]: Too many trailing lines.

File: src/util/coloredstring.cpp [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+
+/*std::wstring ColoredString::serialize() const {
+	std::wstring out;
+	for (size_t i = 0; i < m_string.size(); ++i) {
+		if (i == 0 || m_colors[i] != m_colors[i - 1])
+			out += L"\vAABBCC";
+		out += m_string[i];
+	}
+	return out;
+}*/
```
Comment (ShadowNinja) [inline src/util/coloredstring.cpp]: Remove.

Comment (nerzhul) [inline src/util/coloredstring.cpp]: Please note this PR is incompatible with 0.4 clients and lesser which cannot decode color codes. This PR is reserved to dev-0.5

**Comment by nerzhul** [comment]
Thanks for your comments @ShadowNinja , i hadn't time to work on it yesterday and only fix some directories.
Warning: this change needs 0.5 only because old clients cannot understand the colored strings. \0 could not be used for escape but i will work on it

File: src/chat.cpp (hunk 2 above)
Comment (nerzhul) [inline src/chat.cpp]: I agree.

**Comment by TriBlade9** [comment]
@nerzhul Old clients will just see the escape codes. It wouldn't crash them at least.

**Comment by nerzhul** [comment]
@TriBlade9 yes, i know, this is unacceptable. The messages must be proper, then it will be a 0.5 feature only.

File: src/util/string.cpp (hunk 5 above)
Comment (nerzhul) [inline src/util/string.cpp]: i also notices the out of bounds issue, i'll fix it

**Comment by nerzhul** [comment]
Rebased, @ShadowNinja can you check ?

File: src/client/guiChatConsole.cpp [hunk 12]
```diff
@@ -572,4 +586,3 @@ bool GUIChatConsole::OnEvent(const SEvent& event)
 
 	return Parent ? Parent->OnEvent(event) : false;
 }
-
```
Comment (ShadowNinja) [inline src/client/guiChatConsole.cpp]: There should be one (and only one) trailing line at the end of each file.

File: src/util/string.cpp [hunk 13]
```diff
@@ -613,8 +613,41 @@ static bool parseNamedColorString(const std::string &value, video::SColor &color
...
+
 void str_replace(std::string &str, char from, char to)
 {
 	std::replace(str.begin(), str.end(), from, to);
 }
-
```
Comment (ShadowNinja) [inline src/util/string.cpp]: Should have one and only one trailing line.

**Comment by nerzhul** [comment]
@ShadowNinja fixed

**Comment by ShadowNinja** [comment]
Still missing reset token, named colors, and extensibility...

**Comment by nerzhul** [comment]
for this PR i will only add reset token. The named colors will be done in a future PR.

What do you mean with extensibility ?

**Comment by ShadowNinja** [comment]
@nerzhul: We should be able to, for example, make `"\vb"` toggle bold text, `"\v(italics:on)"` turn on italics, `"\v(link:'http://www.minetest.net'")` add a clickable link, or `"\v#(green)"` add a named color.  I propose that color codes are changed to something like `"\v#0123456"` (assuming you keep the `"\v"` escape).

`NUL` doesn't seem to be properly kept in chat strings (probably a simple `lua_getstring` or `char *` somewhere), but that should be trivially fixable.

I also noticed something else: this doesn't have a vertical tab escape, so you can't send a literal vertical tab (although vertical tabs don't work quite right at the moment, but that could be fixed).

**Comment by nerzhul** [comment]
For this PR i limit to reset token which is the base, i haven't time to write a full parser, more PR can write added later :)

File: builtin/game/misc.lua [hunk 14]
```diff
@@ -112,3 +112,18 @@ function core.record_protection_violation(pos, name)
 	end
 end
 
+function core.color(color)
```
Comment (est31) [inline builtin/game/misc.lua]: Shouldn't function names describe an action?

**Comment by kilbith** [comment]
As many players are expecting that feature and accordingly that @nerzhul did nothing on that PR within one month, is someone willing for taking it in charge and finalizing stuff ? @ShadowNinja, maybe ?

**Comment by t4im** [comment]
linking to #1587 for overview

**Comment by kilbith** [comment]
@est31 That feature is worth for being in your roadmap, paired with the [UTF-8 based chat](https://github.com/minetest/minetest/pull/2885) (if it doesn't conflicts too deeply).

**Comment by kilbith** [comment]
Please someone re-open this.

**Comment by est31** [comment]
@kilbith not possible, this pr was done against the `0.5-dev` branch, which has been removed since.

I think nrz did not intend to close this PR itself, but instead has removed the `0.5-dev` branch, this automatically closed the PR.

**Comment by kilbith** [comment]
Then it should be re-opened on a different PR ?

**Comment by WSDguy2014** [comment]
I would like colored chat, for example color for yellow: 
<player> joined the game
<player> left the game

**Pull request closed without merging**